src/program.rs

```diff
@@ -0,0 +1,315 @@
+            outlives: facts.outlives.into_iter().collect(),
+            region_live_at: facts.region_live_at.into_iter().collect(),
+            invalidates: facts.invalidates.into_iter().collect(),
+        }
+    }
+}
+
+/// Parses an input program into a set of its facts, into the same format `rustc` outputs.
+crate fn parse_from_program(
+    program: &str,
+    tables: &mut InternerTables,
+) -> Result<AllFacts, String> {
+    let input = parse_input(program)?;
+
+    let mut facts: Facts = Default::default();
+
+    // facts: universal_region(Region)
+    for region in &input.universal_regions {
+        let region = tables.regions.intern(region);
+        facts.universal_region.insert(region);
+    }
+
+    for block in &input.blocks {
+        let block_name = &block.name;
+
+        for (statement_idx, statement) in block.statements.iter().enumerate() {
+            let start = format!(
+                "\"Start({block}[{statement}])\"",
+                block = block_name,
+                statement = statement_idx
+            );
+            let mid = format!(
+                "\"Mid({block}[{statement}])\"",
+                block = block_name,
+                statement = statement_idx
+            );
+
+            let start = tables.points.intern(&start);
+            let mid = tables.points.intern(&mid);
+
+            // facts: cfg_edge(Point, Point)
+            {
+                if statement_idx > 0 {
+                    // edge: Previous Mid point to this Start point
+                    let previous_mid = format!(
+                        "\"Mid({block}[{previous_statement}])\"",
+                        block = block_name,
+                        previous_statement = statement_idx - 1
+                    );
+                    let previous_mid = tables.points.intern(&previous_mid);
+
+                    facts.cfg_edge.insert((previous_mid, start));
+                }
+
+                // edge: Start to Mid point
+                facts.cfg_edge.insert((start, mid));
+
+                // goto edges
+                let terminator_idx = block.statements.len() - 1;
+                for goto in &block.goto {
+                    // edge: last Mid point to Start of remote block
+                    let from = format!(
+                        "\"Mid({block}[{statement}])\"",
+                        block = block_name,
+                        statement = terminator_idx
+                    );
+                    let to = format!("\"Start({next_block}[0])\"", next_block = goto);
+
+                    let from = tables.points.intern(&from);
+                    let to = tables.points.intern(&to);
+
+                    facts.cfg_edge.insert((from, to));
+                }
+            }
+
+            // the most common statement effects: mid point effects
+            for effect in &statement.effects {
+                match effect {
+                    // facts: region_live_at(Region, Point)
```

Comment:
Hmm, so for `region_live_at`, we have to do more — we need to actually compute liveness I think, but I realize that to do that we need to also have `def('a)`, of course (and maybe we should make both `def` and `use` accept multiple lifetimes). We could also try to capture the idea of variables with types, though that is creeping towards MIR and I think I had hoped to do that as a second pass.

In other words, as an example:

```
bb0 {
  def('a); // statement 0
  ; // statement 1
  goto bb1;
}

bb1 {
  ; // statement 0
  use('a); // statement 1
}
```

This ought to make `'a` live at all the following points: 

- start(bb0/1), 
- mid(bb0/1), 
- start(bb1/0), 
- mid(bb1/0), 
- start(bb1/1), 
- mid(bb1/1).

The idea would be: some variable `X` whose type includes `'a` is used at mid(bb1/1). So, `'a` is live at that point, because the current value of `X` matters -- this flows backwards until we reach `mid(bb0/0)`, at which point `X` was assigned, so its current value doesn't matter.

Reply:
We can of course use datafrog to do this computation =) you can imagine us emitting `use` and `def` facts and setting up a rule:

```rust
region_live_at(R, P) :- use(R, P).
region_live_at(R, P) :-
  use(R, Q),
  cfg_edge(P, Q),
  !def(R, P).
```

Reply:
We also could potentially refactor the compiler to emit `use` facts and not `region_live_at` facts.